### ISSUE TITLE
Bundle datastore requests

### ISSUE THREAD

**Issue by jsangmeister** [OpenSlides/openslides-autoupdate-service#71]
If you start the main repo and open the agenda, the client issues the following request to the autoupdate service:

    {
        "collection":"meeting",
        "ids":[
            1
        ],
        "fields":{
            "agenda_item_ids":{
                "type":"relation-list",
                "collection":"agenda_item",
                "fields":{
                    "item_number":null,
                    "comment":null,
                    "closed":null,
                    "type":null,
                    "is_hidden":null,
                    "is_internal":null,
                    "duration":null,
                    "weight":null,
                    "level":null,
                    "parent_id":null,
                    "child_ids":null,
                    "meeting_id":null,
                    "content_object_id":{
                        "type":"generic-relation",
                        "fields":{
                            "title":null,
                            "number":null,
                            "list_of_speakers_id":{
                                "type":"relation",
                                "collection":"list_of_speakers",
                                "fields":{
                                    "closed":null,
                                    "speaker_ids":{
                                        "type":"relation-list",
                                        "collection":"speaker",
                                        "fields":{
                                            "begin_time":null,
                                            "end_time":null
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

The autoupdate then issues a huge amount of requests to the datastore which should be reduced. It seems that each model is requested individually, which seems unneccessary. One request per "level" should be enough.

**Comment by ostcar** [comment]
Is this really necessary? There is a cache in the service, so every key is only requested once. If you ran the same request a second time, there should be no database query at all. Even when two users request the same query at the same time, there is only one request the to datastore.

Currently, each model is requested in a separate concourent function. It would not be so easy to bundle the requests.

**Comment by jsangmeister** [comment]
Design-wise it would be the best solution to bundle as much requests as possible to avoid the overhead. If it's too much work, then I'd suggest we'll leave it this way for now, maybe run some performance tests some time to see how much impact it has (probably not much).

**Comment by ostcar** [comment]
I am not sure how the datastore works internally. Does it take the same resources to get one field of a model and to get all fields of a model?

For example, are the this two request equally fast and have the same io and memory usage?

GetMany([{"collection":"agenda_item","ids":[1]}])

GetMany(["agenda_item/1/content_object_id"])

If the datastore gets the full model from postgres anyway I could use the first request to get all fields.

The idea is, that when one client request one field of a model, it is very likely, that this client or another client will request the other fields of the model soon. Therefore they could be requested and saved in the cache.

This would be very easy to implement in the autoupdate service. What do you think about this idea?

**Comment by ostcar** [comment]
For reference. Here are the requests generated with the example data:

```
[meeting/1/agenda_item_ids]
[agenda_item/1/content_object_id]
[agenda_item/15/content_object_id]
[agenda_item/2/content_object_id]
[agenda_item/10/content_object_id]
[agenda_item/8/content_object_id]
[agenda_item/9/content_object_id]
[agenda_item/12/content_object_id]
[agenda_item/14/content_object_id]
[agenda_item/5/content_object_id]
[agenda_item/11/content_object_id]
[agenda_item/13/content_object_id]
[agenda_item/6/content_object_id]
[agenda_item/7/content_object_id]
[agenda_item/3/content_object_id]
[agenda_item/4/content_object_id]
[motion_block/1/list_of_speakers_id]
[motion/1/list_of_speakers_id]
[topic/6/list_of_speakers_id]
[topic/7/list_of_speakers_id]
[motion/2/list_of_speakers_id]
[topic/5/list_of_speakers_id]
[assignment/1/list_of_speakers_id]
[topic/8/list_of_speakers_id]
[motion/4/list_of_speakers_id]
[assignment/2/list_of_speakers_id]
[motion/3/list_of_speakers_id]
[topic/2/list_of_speakers_id]
[topic/1/list_of_speakers_id]
[topic/3/list_of_speakers_id]
[topic/4/list_of_speakers_id]
[list_of_speakers/15/speaker_ids]
[list_of_speakers/1/speaker_ids]
[list_of_speakers/8/speaker_ids]
[list_of_speakers/9/speaker_ids]
[list_of_speakers/2/speaker_ids]
[list_of_speakers/7/speaker_ids]
[list_of_speakers/11/speaker_ids]
[list_of_speakers/10/speaker_ids]
[list_of_speakers/13/speaker_ids]
[list_of_speakers/14/speaker_ids]
[list_of_speakers/12/speaker_ids]
[list_of_speakers/4/speaker_ids]
[list_of_speakers/3/speaker_ids]
[list_of_speakers/5/speaker_ids]
[list_of_speakers/6/speaker_ids]
[agenda_item/1/is_hidden agenda_item/1/is_internal agenda_item/1/duration agenda_item/1/child_ids agenda_item/1/weight agenda_item/15/comment agenda_item/3/comment agenda_item/4/weight agenda_item/5/comment agenda_item/6/type agenda_item/7/type agenda_item/8/item_number agenda_item/9/comment agenda_item/10/closed agenda_item/11/closed agenda_item/12/duration agenda_item/13/comment agenda_item/14/comment agenda_item/2/comment agenda_item/2/closed agenda_item/2/type agenda_item/2/item_number agenda_item/2/parent_id agenda_item/2/is_hidden agenda_item/2/is_internal agenda_item/2/duration agenda_item/2/child_ids agenda_item/2/weight agenda_item/2/level agenda_item/2/meeting_id agenda_item/1/level agenda_item/1/parent_id agenda_item/1/meeting_id agenda_item/1/comment agenda_item/1/closed agenda_item/1/type agenda_item/1/item_number agenda_item/15/closed agenda_item/15/type agenda_item/15/item_number agenda_item/15/weight agenda_item/15/level agenda_item/15/parent_id agenda_item/15/is_hidden agenda_item/15/is_internal agenda_item/15/duration agenda_item/15/child_ids agenda_item/15/meeting_id agenda_item/13/closed agenda_item/10/type agenda_item/11/type agenda_item/12/child_ids agenda_item/9/closed agenda_item/3/closed agenda_item/4/level agenda_item/5/closed agenda_item/6/item_number agenda_item/7/item_number agenda_item/8/comment agenda_item/8/closed agenda_item/8/type agenda_item/8/child_ids agenda_item/14/closed agenda_item/9/type agenda_item/13/type agenda_item/10/item_number agenda_item/10/comment agenda_item/10/is_internal agenda_item/10/duration agenda_item/10/child_ids agenda_item/11/item_number agenda_item/11/comment agenda_item/7/comment agenda_item/3/type agenda_item/4/parent_id agenda_item/5/type agenda_item/6/comment agenda_item/14/type agenda_item/9/item_number agenda_item/3/item_number agenda_item/3/is_hidden agenda_item/13/item_number agenda_item/10/weight agenda_item/9/weight agenda_item/11/is_internal agenda_item/6/closed agenda_item/4/is_hidden agenda_item/5/item_number agenda_item/8/weight agenda_item/12/weight agenda_item/8/level agenda_item/12/level agenda_item/8/parent_id agenda_item/3/is_internal agenda_item/13/weight agenda_item/10/level agenda_item/9/level agenda_item/11/duration agenda_item/14/item_number agenda_item/6/duration agenda_item/4/is_internal agenda_item/5/parent_id agenda_item/12/parent_id agenda_item/12/is_hidden agenda_item/8/is_hidden agenda_item/8/is_internal agenda_item/12/is_internal agenda_item/7/closed agenda_item/12/meeting_id agenda_item/12/type agenda_item/3/duration agenda_item/13/level agenda_item/7/duration agenda_item/10/parent_id agenda_item/10/is_hidden agenda_item/9/parent_id agenda_item/10/meeting_id agenda_item/9/is_hidden agenda_item/11/child_ids agenda_item/14/is_hidden agenda_item/14/is_internal agenda_item/14/duration agenda_item/14/child_ids agenda_item/6/child_ids agenda_item/4/duration agenda_item/5/is_hidden agenda_item/8/duration agenda_item/7/child_ids agenda_item/12/item_number agenda_item/3/child_ids agenda_item/13/parent_id agenda_item/9/is_internal agenda_item/4/child_ids agenda_item/5/is_internal agenda_item/8/meeting_id agenda_item/12/comment agenda_item/3/weight agenda_item/3/level agenda_item/3/parent_id agenda_item/3/meeting_id agenda_item/13/is_hidden agenda_item/13/is_internal agenda_item/13/duration agenda_item/13/child_ids agenda_item/13/meeting_id agenda_item/12/closed agenda_item/5/duration agenda_item/11/weight agenda_item/7/weight agenda_item/9/duration agenda_item/14/weight agenda_item/14/level agenda_item/14/parent_id agenda_item/14/meeting_id agenda_item/5/child_ids agenda_item/5/weight agenda_item/9/child_ids agenda_item/11/level agenda_item/7/level agenda_item/4/meeting_id agenda_item/6/weight agenda_item/6/level agenda_item/6/parent_id agenda_item/6/is_hidden agenda_item/6/is_internal agenda_item/6/meeting_id agenda_item/11/parent_id agenda_item/11/is_hidden agenda_item/11/meeting_id agenda_item/5/level agenda_item/5/meeting_id agenda_item/9/meeting_id agenda_item/7/parent_id agenda_item/7/is_hidden agenda_item/7/is_internal agenda_item/7/meeting_id agenda_item/4/comment agenda_item/4/closed agenda_item/4/type agenda_item/4/item_number motion_block/1/title motion_block/1/number motion/1/title motion/1/number topic/6/title topic/6/number topic/7/title topic/7/number motion/2/title motion/2/number topic/5/title topic/5/number assignment/1/title assignment/1/number topic/8/title topic/8/number motion/4/title motion/4/number assignment/2/title assignment/2/number motion/3/title motion/3/number topic/2/title topic/2/number topic/1/title topic/1/number topic/3/title topic/3/number topic/4/number topic/4/title list_of_speakers/15/closed list_of_speakers/1/closed list_of_speakers/8/closed list_of_speakers/9/closed list_of_speakers/2/closed list_of_speakers/7/closed list_of_speakers/11/closed list_of_speakers/10/closed list_of_speakers/13/closed list_of_speakers/14/closed list_of_speakers/12/closed list_of_speakers/4/closed list_of_speakers/3/closed list_of_speakers/5/closed list_of_speakers/6/closed speaker/13/end_time speaker/13/begin_time speaker/12/begin_time speaker/12/end_time speaker/11/end_time speaker/11/begin_time speaker/5/end_time speaker/5/begin_time speaker/4/end_time speaker/4/begin_time speaker/3/end_time speaker/3/begin_time speaker/8/end_time speaker/8/begin_time speaker/6/end_time speaker/6/begin_time speaker/7/begin_time speaker/7/end_time speaker/10/end_time speaker/10/begin_time speaker/9/end_time speaker/9/begin_time speaker/2/end_time speaker/2/begin_time speaker/1/end_time speaker/1/begin_time]
```

There are 47 requests. 

The current implementation first creates a list of all requested keys and request all keys. It would be possible not to send a request immediately but to wait for a short time (for example 10 Milleseconds). If another key is requested at the same time, both keys are requested together. This would also work with requests from different users. But this would mean, that there is a delay between each level. If I calculate correctly, this example request would take 50 ms longer but the requests would be reduced to 5.

**Comment by jsangmeister** [comment]
> I am not sure how the datastore works internally. Does it take the same resources to get one field of a model and to get all fields of a model?
> 
> For example, are the this two request equally fast and have the same io and memory usage?

No, we explicitely only fetch the requested fields (except when you give a position, but thats not relevant here).

> The idea is, that when one client request one field of a model, it is very likely, that this client or another client will request the other fields of the model soon. Therefore they could be requested and saved in the cache.
> 
> This would be very easy to implement in the autoupdate service. What do you think about this idea?

I don't think that's neccessary, for that, I would just start a new request.


> The current implementation first creates a list of all requested keys and request all keys. It would be possible not to send a request immediately but to wait for a short time (for example 10 Milleseconds). If another key is requested at the same time, both keys are requested together. This would also work with requests from different users. But this would mean, that there is a delay between each level. If I calculate correctly, this example request would take 50 ms longer but the requests would be reduced to 5.

The 47 requests all stem from one request from the client. So, when parsing that in the autoupdate service, you first find that you need the agenda item ids, so you request them. Since you have the full request from the client, you know that you need the content object id of all these items, so you can bundle all next 15 requests into one. And so on for each "layer" of the request.

I don't know how you handle the requests internally, but if you say that's too much work, I'd suggest we'll just leave it that way. Implementing a "waiting timeout" seems like a bad idea and just worse performance-wise than too many requests. Since we have so many connections to the autoupdate service they have to be served ASAP.

**Comment by ostcar** [comment]
I close this for now. I said before, I don't think this will be a problem, thins every fqfield will only be requested once. There will only be requests after an autoupdate instance was started.

If this is a problem we can reopen this issue (or open a new one).

**Comment by ostcar** [comment]
I thought about this and fixed it with #90.

There are now only 5 db-requests. I also hope, that the code of the keysbuilder is now simpler. There is no concurrency anymore.
